fix: gradient calculation error

gongfan99 opened the pull request:
modified:   src/internal/IntersectBSplines.cpp

Hey Dr. Siggel,

First of all, thanks for your generosity for sharing the gordon surface algorithm. I have converted it to a python version ocp_gordon which works very well. Occasionally there is a failure in the intersect point detection. After some debugging, it appears that the gradient calculation in the IntersectBSplines function has a typo. After the correction, the intersect point detection works much more stable now. Hence, I want to contribute back to your code.

I apologize that I do not have the C++ setup so cannot run the tests. But I have checked several times analytically to the gradient calculation.

Thanks again,
Fan Gong

gongfan99 committed:
fix: gradient calculation error
	modified:   src/internal/IntersectBSplines.cpp

rainman110 commented:
Dear Fan Gong, 

Thank you for finding the error and providing the fix. Do you have an example, where the error becomes obvious? 

Probably some b-splines that are not in the range 0...1, right? 

Let me go through the math again by myself, then I'll merge it. 

Thank you very much!

gongfan99 commented:
@rainman110 Below is an example which is not very obvious. I used to have more obvious case but could not find it right now.

These two curves intersects at around the point `(0.5, 0.25, 0)`. Before the correction, it can only converge to the point with a residue error of `7e-05` (the minimizer reports it cannot converge in the end). After the correction, the residue error will be `2.9e-17` and the converge is much quicker.

I believe the reason is that after the bounding box search, the final intersect candidate is boxes.b1.range=[0.4375, 0.5]
boxes.b2.range=[0.0, 1.0] i.e. the first curve has a much narrower parameter span than the second curve which makes the gradient more erroneous.

```json
[
  {
    "poles": [
      [
        0.2,
        0.0,
        0.2
      ],
      [
        0.2827406670180405,
        0.09831137324707118,
        0.13015914662195427
      ],
      [
        0.5119429529674574,
        0.2946196795104728,
        -0.02529557799710079
      ],
      [
        0.8222602601866292,
        0.3825462989760092,
        -0.14683821453466192
      ],
      [
        1.0,
        0.4,
        -0.2
      ]
    ],
    "weights": [
      1.0,
      1.0,
      1.0,
      1.0,
      1.0
    ],
    "knots": [
      0.0,
      0.4397309239946441,
      1.0
    ],
    "mults": [
      4,
      1,
      4
    ],
    "degree": 3,
    "is_periodic": false,
    "first_parameter": 0.0,
    "last_parameter": 1.0
  },
  {
    "poles": [
      [
        -0.0,
        -0.0,
        -0.0
      ],
      [
        1.0,
        0.5,
        -0.0
      ]
    ],
    "weights": [
      1.0,
      1.0
    ],
    "knots": [
      0.0,
      1.0
    ],
    "mults": [
      2,
      2
    ],
    "degree": 1,
    "is_periodic": false,
    "first_parameter": 0.0,
    "last_parameter": 1.0
  }
]
```

rainman110 commented on src/internal/IntersectBSplines.cpp:
```diff
@@ -311,8 +311,8 @@ namespace
-            G(1) = 2. * diff.Dot(d1)* (m_c1->LastParameter()-m_c1->FirstParameter()) * d_getUParam(X.Value(1));
-            G(2) = -2. * diff.Dot(d2) * (m_c2->LastParameter()-m_c2->FirstParameter()) * d_getUParam(X.Value(2));
+            G(1) = 2. * diff.Dot(d1) * d_getUParam(X.Value(1));
+            G(2) = -2. * diff.Dot(d2) * d_getUParam(X.Value(2));
```
```suggestion
            G(2) = -2. * diff.Dot(d2) * d_getVParam(X.Value(2));
```

Hi @gongfan99 . I think it must be also d_getVParam, right? Here we are computing the derivative to v (the second parameter). Can you please verify. I think you are right regarding the removal  of the (last_parm- first_parm) factor.

gongfan99 replied:
Yes, you are correct.

rainman110 committed:
Fixed another gradient issue.

Also added numerical gradient check. With these
changes, the gradient seems to be valid.

rainman110 commented:
@gongfan99 I added a gradient check to the PR. Now the gradient seems to be valid. Note, that indeed the second line of the gradient computation G(2) = ... needs to use d_getVParam !!!

rainman110 committed:
Extended changelog
